Update ticker to map closely to hardware

c1728p9 opened the pull request:
Allow tickers to specify their native frequency and number of bits. This allows the conversion to happen in common code rather than in each vendor's implementation.

c1728p9 commented:
/morph test

c1728p9 commented:
/morph test-nightly

0xc0170 commented:
Please look at failures, tickers mostly

> Allow tickers to specify their native frequency and number of bits. This allows the conversion to happen in common code rather than in each vendor's implementation.

👍  for the info structure addition, still pending reviews

c1728p9 commented:
PR #5096 is needed before testing will pass on this PR. This is because `us_ticker_read()` and `lp_ticker_read()` no longer returns the same value as `ticker_read(ticker)`. The call to `ticker_read(ticker)` both scales the frequency and starts counting from 0.

c1728p9 commented on hal/mbed_ticker_api.c:
```diff
@@ -33,9 +34,31 @@ static void initialize(const ticker_data_t *ticker)
+    uint32_t frequency = info->frequency;
+    if (info->frequency == 0) {
+        MBED_ASSERT(0);
+        frequency = 1;
```
I'll update this to default to 1MHz rather than 1Hz.

c1728p9 commented on hal/mbed_ticker_api.c:
```diff
@@ -33,9 +34,31 @@ static void initialize(const ticker_data_t *ticker)
+        MBED_ASSERT(0);
+        bits = 32;
+    }
+    uint32_t max_delta = 0x7 << (bits - 4); // 7/16th
```
done

c1728p9 commented on hal/mbed_ticker_api.c:
```diff
@@ -86,53 +109,143 @@ static us_timestamp_t convert_timestamp(us_timestamp_t ref, timestamp_t timestam
+        return;
+    }
+
+    uint64_t new_ticks = (ticker_time - queue->tick_last_read) & queue->bitmask;
```
sound good. I updated the variables

c1728p9 commented:
Addressed @pan-'s comments and rebased to master

c1728p9 commented on hal/ticker_api.h:
```diff
@@ -61,13 +64,21 @@ typedef struct {
+    uint64_t max_delta_us;              /**< Largest delta in us that can be used when scheduling */
+    uint32_t tick_last_read;            /**< Last tick read */
+    uint64_t tick_remainder;            /**< Ticks that have not been added to base_time */
+    us_timestamp_t base_time;
```
This wasn't actually being used so I removed it. I also updated the comments next to max_delta to indicate that the delta was in ticks.

c1728p9 commented:
> Should we add here now that deep sleep locking info?

This PR is big enough as it is, it might be best as a separate PR on top of this. What do you think @0xc0170  and @pan-?

pan- commented:
separate PR.

jiandeng commented on hal/mbed_ticker_api.c:
```diff
@@ -86,53 +109,103 @@ static us_timestamp_t convert_timestamp(us_timestamp_t ref, timestamp_t timestam
+    if (match_tick > prev_tick) {
+        return (cur_tick >= match_tick) || (cur_tick < prev_tick);
+    } else {
+        return (cur_tick < match_tick) && (cur_tick >= prev_tick);
```
If match_tick <= prev_tick,  how could prev_tick <= cur_tick < match_tick?

c1728p9 replied:
Good catch! This is a bug.

c1728p9 commented:
Rebased to master and fixed ticker mismatch calculation. Diff for the new changes can be found  in https://github.com/ARMmbed/mbed-os/pull/5028/commits/67127182ab34d7d274ec94b6ef5aa511355cb719

c1728p9 commented:
Rebased to master and squashed bugfix commit

c1728p9 commented:
/morph test-nightly

0xc0170 commented:
Besides couple of nightly runs, I reviewed few, `ticker`  or `lp_timeout` fail, please review the latest test results

c1728p9 commented:
This PR won't pass testing without #5242

theotherjimmy commented:
@c1728p9 It looks like this needs a rebase.

c1728p9 commented:
rebased onto master

0xc0170 commented:
cc @bulislaw @scartmell-arm

0xc0170 commented:
/morph test-nightly

theotherjimmy commented:
retest uvisor

0xc0170 commented:
It is now ready for integration.

@c1728p9 This is backward compatible, thus not required to be minor, is that correct?

c1728p9 commented:
Hi @0xc0170, yes this is backwards compatible.

theotherjimmy commented:
@AnotherButler More new APIs here.

AnotherButler commented:
@theotherjimmy Thanks. I've made a note of that and talked to @c1728p9 about docs. This should not affect this PR being merged.
